ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java

```diff
@@ -352,24 +371,35 @@ protected boolean shouldReplicateFromDc(String datacenterName) {
    * 1. On a node become online role of partition, which usually happens on restart or deployment.
    * 2. On Ambry cluster change.
    */
-  private void scheduleVcrHelix() {
+  private void scheduleVcrHelix(String reason) {
     if (vcrHelixUpdateFuture != null && vcrHelixUpdateFuture.cancel(false)) {
       // If a vcrHelixUpdate task is scheduled, try to cancel it first.
       logger.info("There was a scheduled vcrHelixUpdate task. Canceled.");
       vcrHelixUpdateFuture = null;
     }
     // either success cancel or not, we should schedule a new job to updateVcrHelix
     vcrHelixUpdateFuture =
-        scheduler.schedule(() -> updateVcrHelix(), cloudConfig.vcrHelixUpdateDelayTimeInSeconds, TimeUnit.SECONDS);
+        scheduler.schedule(() -> updateVcrHelix(reason), cloudConfig.vcrHelixUpdateDelayTimeInSeconds,
+            TimeUnit.SECONDS);
     logger.info("VcrHelixUpdate task scheduled. Will run in {} seconds.", cloudConfig.vcrHelixUpdateDelayTimeInSeconds);
   }
 
   /**
    * The actual performer to update VCR Helix:
    */
-  synchronized private void updateVcrHelix() {
-    logger.info("Going to update VCR Helix Cluster. Dryrun: {}", cloudConfig.vcrHelixUpdateDryRun);
-    logger.info("Current partitions in clustermap data structure: {}",
+  synchronized private void updateVcrHelix(String reason) {
+
+    logger.info("Going to update VCR Helix Cluster. Reason: {}, Dryrun: {}", reason, cloudConfig.vcrHelixUpdateDryRun);
+    while (!vcrUpdateDistributedLock.tryLock()) {
+      logger.warn("Could not obtain vcr update distributed lock. Sleep and retry.");
```

Comment:
The only way for a vcr node to get out of this while loop, is to have helix assign parttionid 1 to it, right?
What if that never happens., and helix never assigns partition 1 to this node. The vcr process will have to be restarted to get out of this loop. Should we limit the number of times sleep can happen in line 396? Something like, if we have slept and tried for lock 5 times, then its better to give up, because something else might be wrong. If we miss updating the cluster somehow, then anyways we will get notification due to alerts on the metric that checks if the cluster are in sync.

Reply:
the lock is not associated to partition 1. It's a general distributed lock on zk. It will eventually get the lock no matter if the node has partition 1 or not. 

I agree with the retry count concern. I can add a config for max retry count.